File: frontend/src/api/community.ts
import apiClient from "./client";

export interface Building {
  id: string;
  name: string;
  address: string | null;
  invite_code: string | null;
  created_at: string;
}

export interface CommunityBook {
  id: string;
  owner_id: string;
  owner_name: string;
  owner_avatar: string | null;
  catalog_id: string;
  title: string;
  authors: string[] | null;
  cover_url: string | null;
  deposit_amount: number;
  lend_note: string | null;
  is_blocked: boolean;
}

export interface Member {
  id: string;
  name: string;
  avatar_url: string | null;
  profile_slug: string | null;
  phone_verified: boolean;
}

export const communityApi = {
  createBuilding: (name: string, address?: string) =>
    apiClient.post<Building>("/buildings", { name, address }),

  join: (invite_code: string) =>
    apiClient.post<Building>("/buildings/join", { invite_code }),

  myBuilding: () => apiClient.get<Building>("/buildings/me"),

  books: (params?: { limit?: number; offset?: number }) =>
    apiClient.get<CommunityBook[]>("/buildings/books", { params }),

  members: () => apiClient.get<Member[]>("/buildings/members"),
};

File: backend/app/schemas/community.py
from __future__ import annotations

import uuid
from datetime import datetime
from typing import Optional

from pydantic import BaseModel, ConfigDict


class BuildingCreate(BaseModel):
    name: str
    address: Optional[str] = None


class BuildingJoin(BaseModel):
    invite_code: str


class BuildingOut(BaseModel):
    model_config = ConfigDict(from_attributes=True)

    id: uuid.UUID
    name: str
    address: Optional[str] = None
    invite_code: Optional[str] = None
    created_at: datetime


class MemberOut(BaseModel):
    model_config = ConfigDict(from_attributes=True)

    id: uuid.UUID
    name: str
    avatar_url: Optional[str] = None
    profile_slug: Optional[str] = None
    phone_verified: bool


class CommunityBookOut(BaseModel):
    """A user_book available for borrowing in the community."""
    model_config = ConfigDict(from_attributes=True)

    id: uuid.UUID
    owner_id: uuid.UUID
    owner_name: str
    owner_avatar: Optional[str]
    catalog_id: uuid.UUID
    title: str
    authors: Optional[list[str]]
    cover_url: Optional[str]
    deposit_amount: float
    lend_note: Optional[str]
    is_blocked: bool  # True if current_user is blacklisted by this lender
